lib/src/extend/functions.dart

```diff
@@ -106,19 +104,63 @@ List<ComplexSelector>? unifyComplex(
 /// Returns a [CompoundSelector] that matches only elements that are matched by
 /// both [compound1] and [compound2].
 ///
-/// The [span] will be used for the new unified selector.
+/// The [compound1]'s `span` will be used for the new unified selector.
+///
+/// This function ensures that the relative order of pseudo-classes (`:`) and
+/// pseudo-elements (`::`) within each input selector is preserved in the
+/// resulting combined selector.
+///
+/// This function enforces a general rule that pseudo-classes (`:`) should come
+/// before pseudo-elements (`::`), but it won't change their order if they were
+/// originally interleaved within a single input selector. This prevents
+/// unintended changes to the selector's meaning. For example, unifying
+/// `::foo:bar` and `:baz` results in `:baz::foo:bar`. `:baz` is a pseudo-class,
+/// so it is moved before the pseudo-class `::foo`. Meanwhile, `:bar` is not
+/// moved before `::foo` because it appeared after `::foo` in the original
+/// selector.
 ///
 /// If no such selector can be produced, returns `null`.
 CompoundSelector? unifyCompound(
     CompoundSelector compound1, CompoundSelector compound2) {
-  var result = compound2.components;
-  for (var simple in compound1.components) {
-    var unified = simple.unify(result);
-    if (unified == null) return null;
-    result = unified;
+  var result = <SimpleSelector>[];
+  var pseudoResult = <SimpleSelector>[];
+  var pseudoElementFound = false;
+
+  // Compound selectors may not contain more than one pseudo-element selectors.
+  if (compound1.components
+          .followedBy(compound2.components)
+          .whereType<PseudoSelector>()
+          .where((pseudo) => pseudo.isElement)
+          .toSet()
+          .length >
+      1) {
```

Comment:
Unification can be part of a tight loop in `@extend`-heavy stylesheets, at which point allocating a bunch of new objects gets expensive. Consider checking whether both compound selectors even have pseudo-elements at all before doing this check.

Reply:
Does allocation happen with `toSet()`? IIUC the other functions are lazy. With that in mind, I'll check for the `length` being > 1 in the first place before doing `toSet()`. Does that sound reasonable?

Reply:
actually, since I removed the tests where one selector contains `::x::y`, then I think we don't need this. The consequence is that `selector.unify('::x::y', 'z')` will produce `z::x::y` instead of `null`, but the input selectors don't make sense.. so I think that's OK?

Reply:
Yeah, that's fine. That's the sort of thing we'd only really expect to be accurate if we started parsing pseudo-elements as combinators.